File: objc/test/objc/Ice/dispatcher/AllTests.m
// **********************************************************************
//
// Copyright (c) 2003-2011 ZeroC, Inc. All rights reserved.
//
// This copy of Ice Touch is licensed to you under the terms described in the
// ICE_TOUCH_LICENSE file included in this distribution.
//
// **********************************************************************

#import <Ice/Ice.h>
#import <TestCommon.h>
#import <DispatcherTest.h>

#import <Foundation/Foundation.h>

static BOOL isDispatcherThread()
{
    return dispatch_get_current_queue() == dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0);
}

@interface TestDispatcherCallback : NSObject
{
    BOOL called;
    NSCondition* cond;
}
-(void) check;
-(void) called;
@end

@implementation TestDispatcherCallback
-(id) init
{
    if(![super init])
    {
        return nil;
    }
    cond = [[NSCondition alloc] init];
    return self;
}
-(void) dealloc
{
    [cond release];
    [super dealloc];
}
-(void) check
{
    [cond lock];
    while(!called)
    {
        [cond wait];
    }
    called = NO;
    [cond unlock];
}
-(void) called
{
    [cond lock];
    called = YES;
    [cond signal];
    [cond unlock];
}
-(void) response
{
    test(isDispatcherThread());
    [self called];
}

-(void) exception:(ICEException*)ex
{
    test([ex isKindOfClass:[ICENoEndpointException class]]);
    test(isDispatcherThread());
    [self called];
}

-(void) payload
{
    test(isDispatcherThread());
}

-(void) ignoreEx:(ICEException*)ex
{
    test([ex isKindOfClass:[ICECommunicatorDestroyedException class]]);
}

-(void) sent:(BOOL)sentSynchronously
{
    test(sentSynchronously || isDispatcherThread());
}

@end

id<TestDispatcherTestIntfPrx>
dispatcherAllTests(id<ICECommunicator> communicator)
{
    NSString* sref = @"test:default -p 12010";
    id<ICEObjectPrx> obj = [communicator stringToProxy:sref];
    test(obj);

    id<TestDispatcherTestIntfPrx> p = [TestDispatcherTestIntfPrx uncheckedCast:obj];

    sref = @"testController:tcp -p 12011";
    obj = [communicator stringToProxy:sref];
    test(obj);

    id<TestDispatcherTestIntfControllerPrx> testController = [TestDispatcherTestIntfControllerPrx uncheckedCast:obj];

    tprintf("testing dispatcher... ");
    {
        [p op];

        TestDispatcherCallback* cb = [[TestDispatcherCallback alloc] init];
        [p begin_op:^ { [cb response]; } exception:^(ICEException* ex) { [cb exception:ex]; }];
        [cb check];

        TestDispatcherTestIntfPrx* i = [p ice_adapterId:@"dummy"];
        [i begin_op:^ { [cb response]; } exception:^(ICEException* ex) { [cb exception:ex]; }];
        [cb check];

        [testController holdAdapter];

        ICEByte d[1024];
        ICEMutableByteSeq* seq = [ICEMutableByteSeq dataWithBytes:d length:sizeof(d)];
        id<ICEAsyncResult> result;
        while([(result = [p begin_opWithPayload:seq response:^{ [cb payload]; } 
                                      exception:^(ICEException* ex) { [cb ignoreEx:ex]; }
                                           sent:^(BOOL ss) { [cb sent:ss]; }]) sentSynchronously]);
        [testController resumeAdapter];
        [result waitForCompleted];

        [cb release];
    }
    tprintf("ok\n");

    return p;
}

File: objc/test/cpp/Ice/dispatcher/AllTests.cpp
// **********************************************************************
//
// Copyright (c) 2003-2012 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_TOUCH_LICENSE file included in this distribution.
//
// **********************************************************************

#include <Ice/Ice.h>
#include <IceUtil/Random.h>
#include <TestCommon.h>
#include <DispatcherTest.h>
#include <dispatcher/Dispatcher.h>

using namespace std;
using namespace Test::Dispatcher;

namespace
{
    
    class Callback : public IceUtil::Shared
    {
    public:
        
        Callback() :
        _called(false)
        {
        }
        
        void check()
        {
            IceUtil::Monitor<IceUtil::Mutex>::Lock sync(_m);
            while(!_called)
            {
                _m.wait();
            }
            _called = false;
        }
        
        void
        response()
        {
            test(ClientDispatcher::isDispatcherThread());
            called();
        }
        
        void
        exception(const Ice::Exception& ex)
        {
            test(dynamic_cast<const Ice::NoEndpointException*>(&ex));
            test(ClientDispatcher::isDispatcherThread());
            called();
        }
        
        void
        payload()
        {
            test(ClientDispatcher::isDispatcherThread());
        }
        
        void
        ignoreEx(const Ice::Exception& ex)
        {
            test(dynamic_cast<const Ice::CommunicatorDestroyedException*>(&ex));
        }
        
        void
        sent(bool sentSynchronously)
        {
            test(sentSynchronously || ClientDispatcher::isDispatcherThread());
        }
        
    protected:
        
        void called()
        {
            IceUtil::Monitor<IceUtil::Mutex>::Lock sync(_m);
            assert(!_called);
            _called = true;
            _m.notify();
        }
        
    private:
        
        IceUtil::Monitor<IceUtil::Mutex> _m;
        bool _called;
    };
    typedef IceUtil::Handle<Callback> CallbackPtr;
    
}

void
dispatcherAllTests(const Ice::CommunicatorPtr& communicator)
{
    string sref = "test:default -p 12010";
    Ice::ObjectPrx obj = communicator->stringToProxy(sref);
    test(obj);
    
    Test::Dispatcher::TestIntfPrx p = Test::Dispatcher::TestIntfPrx::uncheckedCast(obj);
    
    sref = "testController:tcp -p 12011";
    obj = communicator->stringToProxy(sref);
    test(obj);
    
    Test::Dispatcher::TestIntfControllerPrx testController = Test::Dispatcher::TestIntfControllerPrx::uncheckedCast(obj);
    
    tprintf("testing dispatcher... ");
    {
        p->op();
        
        CallbackPtr cb = new Callback;
        Test::Dispatcher::Callback_TestIntf_opPtr callback = Test::Dispatcher::newCallback_TestIntf_op(cb,
                                                                               &Callback::response,
                                                                               &Callback::exception);
        p->begin_op(callback);
        cb->check();
        
        Test::Dispatcher::TestIntfPrx i = p->ice_adapterId("dummy");
        i->begin_op(callback);
        cb->check();
        
        testController->holdAdapter();
        
        Test::Dispatcher::Callback_TestIntf_opWithPayloadPtr callback2 =
        Test::Dispatcher::newCallback_TestIntf_opWithPayload(cb, &Callback::payload, &Callback::ignoreEx, &Callback::sent);
        
        Ice::ByteSeq seq;
        seq.resize(1024); // Make sure the request doesn't compress too well.
        for(Ice::ByteSeq::iterator q = seq.begin(); q != seq.end(); ++q)
        {
            *q = static_cast<Ice::Byte>(IceUtilInternal::random(255));
        }
        Ice::AsyncResultPtr result;
        while((result = p->begin_opWithPayload(seq, callback2))->sentSynchronously());
        testController->resumeAdapter();
        result->waitForCompleted();
    }
    tprintf("ok\n");
    
    p->shutdown();
}
